encoding.go

```diff
@@ -71,74 +100,76 @@ func ParseStatelessNode(serialized []byte, depth byte, comm []byte) (VerkleNode,
 			}
-			values[i] = serialized[offset : offset+32]
-			offset += 32
+			values[i] = serialized[offset : offset+LeafValueSize]
+			offset += LeafValueSize
 		}
 	}
-	if NodeWidth != len(values) {
-		return nil, fmt.Errorf("invalid number of nodes in decoded child expected %d, got %d", NodeWidth, len(values))
-	}
-	ln := NewLeafNodeWithNoComms(serialized[1:32], values[:])
+	ln := NewLeafNodeWithNoComms(serialized[leafSteamOffset:leafSteamOffset+StemSize], values[:])
 	ln.setDepth(depth)
 	ln.c1 = new(Point)
-	ln.c1.SetBytesTrusted(serialized[64:96])
+	ln.c1.SetBytesUncompressed(serialized[leafC1CommitmentOffset:leafC1CommitmentOffset+SerializedPointSize], true)
 	ln.c2 = new(Point)
-	ln.c2.SetBytesTrusted(serialized[96:128])
+	ln.c2.SetBytesUncompressed(serialized[leafC2CommitmentOffset:leafC2CommitmentOffset+SerializedPointSize], true)
 	ln.commitment = new(Point)
-	ln.commitment.SetBytesTrusted(comm)
+	ln.commitment.SetBytes(comm, true)
 	return ln, nil
 }
 
-func deserializeIntoStateless(bitlist []byte, raw []byte, depth byte, comm []byte) (*StatelessNode, error) {
+func deserializeIntoStateless(bitlist []byte, raw []byte, depth byte, comm SerializedPointCompressed) (*StatelessNode, error) {
 	// GetTreeConfig caches computation result, hence
 	// this op has low overhead
 	n := NewStateless()
 	n.setDepth(depth)
 	indices := indicesFromBitlist(bitlist)
-	if len(raw)/32 != len(indices) {
+	if len(raw)/SerializedPointSize != len(indices) {
 		return nil, ErrInvalidNodeEncoding
 	}
 	for i, index := range indices {
-		n.unresolved[byte(index)] = raw[i*32 : (i+1)*32]
+		n.unresolved[byte(index)] = raw[i*SerializedPointSize : (i+1)*SerializedPointSize]
 	}
 	n.commitment = new(Point)
-	n.commitment.SetBytesTrusted(comm)
+	n.commitment.SetBytes(comm, true)
 	return n, nil
 }
 
-func CreateInternalNode(bitlist []byte, raw []byte, depth byte, comm []byte) (*InternalNode, error) {
+func CreateInternalNode(bitlist []byte, raw []byte, depth byte, comm SerializedPointCompressed) (*InternalNode, error) {
 	// GetTreeConfig caches computation result, hence
 	// this op has low overhead
 	n := (newInternalNode(depth)).(*InternalNode)
 	indices := indicesFromBitlist(bitlist)
-	if len(raw)/32 != len(indices) {
+
+	if len(raw)/SerializedPointSize != len(indices) {
 		return nil, ErrInvalidNodeEncoding
 	}
+
+	freelist := make([]HashedNode, len(indices))
 	for i, index := range indices {
-		n.children[index] = &HashedNode{commitment: raw[i*32 : (i+1)*32]}
+		freelist[i].serialized = raw[i*SerializedPointSize : (i+1)*SerializedPointSize]
+		n.children[index] = &freelist[i]
 	}
 	n.commitment = new(Point)
-	n.commitment.SetBytesTrusted(comm)
+	n.commitment.SetBytes(comm, true)
 	return n, nil
 }
 
 func indicesFromBitlist(bitlist []byte) []int {
-	indices := make([]int, 0)
+	indices := make([]int, 0, 32)
```

Comment:
This obvious optimization saved a ton of allocations and GC pressure, which looks obvious in hindsight.
We can only have up to 32 indices, so we do a single allocation with that capacity.

This `indicesFromBitlist(...)` was called _a lot_ of times in the replay benchmark from `go-ethereum`.